Proofs/Experiments/AbsVCG.lean

```diff
@@ -0,0 +1,333 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Shilpi Goel
+
+Experimental: Use the Correctness module to prove that this program implements
+absolute value correctly. The objective of this exercise is to determine the
+"shape" of the lemmas that are needed for automation and/or to modify the
+definitions in Correctness, if needed.
+-/
+import Arm
+import Tactics.StepThms
+import Tactics.Sym
+import Correctness.ArmSpec
+import Lean
+
+namespace AbsVCG
+
+def program : Program :=
+  def_program
+   [
+    (0x4005d0#64, 0x2a0003e1#32), --  mov w1, w0
+    (0x4005d4#64, 0x131f7c00#32), --  asr w0, w0, #31
+    (0x4005d8#64, 0x0b000021#32), --  add w1, w1, w0
+    (0x4005dc#64, 0x4a000020#32), --  eor w0, w1, w0
+    (0x4005e0#64, 0xd65f03c0#32)] --  ret
+
+/-- Precondition for the correctness of the `Abs` program. -/
+def abs_pre (s : ArmState) : Prop :=
+  read_pc s = 0x4005d0#64 ∧
+  s.program = program ∧
+  read_err s = StateError.None
+
+/-- Specification of the absolute value computation for a 32-bit bitvector. -/
+def spec (x : BitVec 32) : BitVec 32 :=
+  -- We prefer the current definition as opposed to:
+  -- BitVec.ofNat 32 x.toInt.natAbs
+  -- because the above has functions like `toInt` that do not play well with
+  -- bitblasting/LeanSAT.
+  let msb := BitVec.extractLsb 31 31 x
+  if msb == 0#1 then
+    x
+  else
+    (0#32 - x)
```

Comment:
It should be doable to prove this definition equal to `x.toInt.natAbs` by rewriting along `[toInt_eq_msb_cond]`(https://leanprover-community.github.io/mathlib4_docs/Init/Data/BitVec/Lemmas.html#BitVec.toInt_eq_msb_cond).

That said, for this experiment it makes total sense not to bother with that and just use the more convenient definition you have here